packages/lit-dev-server/src/index.ts

```diff
@@ -43,28 +43,21 @@ if (mode === 'playground') {
 }
 
 app.use(async (ctx, next) => {
-  if (ctx.path.match(/\/[^\/\.]+$/)) {
+  let path = ctx.path;
+  if (path.match(/\/[^\/\.]+$/)) {
     // Canonicalize paths to have a trailing slash, except for files with
     // extensions.
-    ctx.status = 301;
-    ctx.redirect(
-      ctx.path + '/' + (ctx.querystring ? '?' + ctx.querystring : '')
-    );
-  } else if (ctx.path.endsWith('//')) {
+    path += '/';
+  }
+  if (path.endsWith('//')) {
     // Koa static doesn't care if there are any number of trailing slashes.
     // Normalize this too.
+    path = path.replace(/\/+$/, '/');
+  }
+  path = pageRedirects.get(path) ?? path;
+  if (path !== ctx.path) {
```

Comment:
Elegant!

Especially like how this check handles the three different `301` cases.

Reply:
Nice, but... can we add a comment? :)

Reply:
Added a comment about how it's to flatten redirect chains so we don't actually serve a chain of them.